vonk: Add documentation about upgrading Vonk. (#28)
simplifhir/simplifier-docs: vonk/releasenotes/upgrade.rst

   #. If needed, an update is applied to the database structure.
   #. If Vonk introduces a new version of the FHIR .NET API, Vonk will load a new set of Conformance Resources from the specification.zip into the Administration database, for both FHIR STU3 and FHIR R4. In a specific case you can :ref:`prevent this step from happening <replace_admindb>`.

#. When Vonk is done with the tasks above, it is again available to process requests.
#. Check the log for warnings stating that you use obsolete settings. If so, adjust them and restart Vonk.

If anything went wrong, go back:

#. Stop the (new) running instance of Vonk.
#. Restore both databases from your backup.
#. Switch to the old installation directory and start the old version of Vonk from there (``> dotnet .\Vonk.Server.dll``)
#. It should start as it did before you began the upgrade.
#. Report the problem to the Vonk helpdesk, see :ref:`vonk-contact`.

.. _replace_admindb:

You may be able to avoid the import of specification.zip if:

* The Administration database is in SQLite and
* You have not made alterations to the Administration API through the Web API.

> In this case you can simply replace the old database (usually with the filename vonkadmin.db) with the one from the new installation directory (in ./data/vonkadmin.db).

Reviewer: ```suggestion
In this case you can simply replace the old database (usually with the filename ``vonkadmin.db``) with the one from the new installation directory (in ``./data/vonkadmin.db``).
```

Do so *before* you start the new Vonk installation.
Anything specified in :ref:`AdministrationImportOptions <configure_admin_import>` will be re-imported into the new database.

.. _upgrade_server_docker:

Using Docker
^^^^^^^^^^^^

Revisit :ref:`use_docker`.

#. Stop the running container for Vonk: ``> docker stop vonk.server``.
#. Pull the latest image for Vonk: ``> docker pull simplifier/vonk``
#. Check the :ref:`vonk_releasenotes` for any new settings that you may want to apply or change from their defaults, and apply that to the ``environment`` setting in the docker-compose file.
#. Check the :ref:`vonk_releasenotes` for any action that you need to take specifically for this upgrade.

> #. Create a backup of your current databases, both the main Resource database and the Administration database. (See :ref:`configure_repository` and your docker-compose file to find the details on where your databases are).

Reviewer: ```suggestion
#. Create a backup of your current databases, both the main Resource database and the Administration database. See :ref:`configure_repository` and your docker-compose file to find the details on where your databases are.
```

#. Vonk will now do several upgrade tasks, during which any web request will be responded to with 423 - Locked:

   #. If needed, an update is applied to the database structure.
   #. If Vonk introduces a new version of the FHIR .NET API, Vonk will load a new set of Conformance Resources from the specification.zip into the Administration database, for both FHIR STU3 and FHIR R4. In a specific case you can :ref:`prevent this step from happening <replace_admindb>`.

#. When Vonk is done with the tasks above, it is again available to process requests.
#. Check the log for warnings stating that you use obsolete settings. If so, adjust them and restart Vonk.

If anything went wrong, go back:

#. Stop the (new) running container of Vonk.
#. Restore both databases from your backup.
#. Specify your previous image of Vonk in the docker command or in the docker-compose file: ``simplifier\vonk:<previous-version-tag>``
#. Start the container based on this previous image.
#. It should start as it did before you began the upgrade.
#. Report the problem to the Vonk helpdesk, see :ref:`vonk-contact`.

.. _upgrade_plugin:

Upgrading Plugins
-----------------

> Since a Plugin runs in the context of a Vonk FHIR Server we advice you to start by upgrading your Vonk FHIR Server, without loading your Plugin.

Reviewer: ```suggestion
Since a Plugin runs in the context of a Vonk FHIR Server we advise you to start by upgrading your Vonk FHIR Server, without loading your Plugin.
```

Check the section on :ref:`settings_pipeline` to see how you can exclude your plugin from the pipeline.

Upgrade the references in your plugin:

> #. Open the sourcecode of your plugin, and open the project file (``yourplugin.csproj``).

Reviewer: ```suggestion
#. Open the source code of your plugin, and open the project file (``yourplugin.csproj``).
```
